cloud-controller-manager/do/kubernetes_lb_validator_test.go

```diff
@@ -0,0 +1,271 @@
+				URL:    &url.URL{},
+			},
+			StatusCode: http.StatusUnprocessableEntity,
+			Body:       ioutil.NopCloser(bytes.NewBufferString("test")),
+		},
+	}
+
+}
+
+func Test_Handle(t *testing.T) {
+	os.Setenv(regionEnv, "nyc3")
+
+	testcases := []struct {
+		name            string
+		req             admission.Request
+		gCLient         *godo.Client
+		expectedAllowed bool
+		resp            *godo.Response
+		err             error
+	}{
+		{
+			name: "Allow if service type is not load balancer",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: "test",
+					},
+					Spec: corev1.ServiceSpec{
+						Type: corev1.ServiceTypeNodePort,
+					},
+				},
+				nil,
+			)},
+			expectedAllowed: true,
+		},
+		{
+			name: "Allow if request is of type DELETE",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: "test",
+						DeletionTimestamp: &metav1.Time{
+							Time: time.Now().UTC(),
+						},
+					},
+					Spec: corev1.ServiceSpec{
+						Type: corev1.ServiceTypeLoadBalancer,
+					},
+				},
+				nil,
+			)},
+			expectedAllowed: true,
+		},
+		{
+			name: "Allow CREATE happy path",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: "test",
+					},
+					Spec: corev1.ServiceSpec{
+						Type: corev1.ServiceTypeLoadBalancer,
+					},
+				},
+				nil,
+			)},
+			expectedAllowed: true,
+		},
+		{
+			name: "Deny CREATE invalid configuration",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "test",
+						Namespace: "test",
+					},
+					Spec: corev1.ServiceSpec{
+						Type: corev1.ServiceTypeLoadBalancer,
+					},
+				},
+				nil,
+			)},
+			resp:            newFakeUnprocessableResponse(),
+			err:             newFakeUnprocessableErrorResponse(),
+			expectedAllowed: false,
+		},
+		{
+			name: "Allow Update happy path",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2"), fakeService("old-service"))},
+			expectedAllowed: true,
+		},
+		{
+			name: "Deny Update invalid configuration",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2"), fakeService("old-service"))},
+			expectedAllowed: false,
+			resp:            newFakeUnprocessableResponse(),
+			err:             newFakeUnprocessableErrorResponse(),
+		},
+		{
+			name: "Deny Update validation error",
+			req: admission.Request{AdmissionRequest: fakeAdmissionRequest(
+				fakeService("test2"), fakeService("old-service"))},
+			expectedAllowed: false,
+			resp:            newFakeUnprocessableResponse(),
+			err:             newFakeUnprocessableErrorResponse(),
```

Comment:
I don't think `resp` and `err` here can be same as in the `"Deny Update invalid configuration"` sub-test: AFAIU, they should yield different responses (errored vs denied), and so the faked response/error must be different too?

(This is probably also a good example why checking for exact/substring errors is important.)

Reply:
@timoreimann Errored response doesn't return a message like alllow and deny do. It does however return the passed in status code while the others return 200 or 403. I've added an assertion on the status code

Reply:
Aren't we returning a `admission.Errored()` response when we hit any kind of (non-validation) error?

This returns a `Response` struct (just like `admission.Allowed()` and `admission.Denied()` do) which embeds the passed in error [here](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/a69e9c3b138c092f51787fcb90a796d33eccc1af/vendor/sigs.k8s.io/controller-runtime/pkg/webhook/admission/response.go#L56). This is what we should be able to assert on, thereby allowing us to distinguish between different error scenarios reliably (because the response code alone wouldn't be sufficient in all cases).